doc/lp5250d.1
'\" t
.ig
Man page for lp5250d.

Copyright (C) 1999 - 2008 Carey Evans.

You can redistribute and/or modify this document under the terms of 
the GNU General Public License as published by the Free Software
Foundation; either version 2 of the License, or (at your option)
any later version.

This document is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.
..
.TH LP5250D 1 "12 November 2001"
.SH NAME
lp5250d \- start an AS/400 printer session
.SH SYNOPSIS
.B lp5250d
.RI [\| OPTIONS \|]
.I HOSTNAME
.SH "DESCRIPTION"
.B lp5250d
connects to an AS/400 over TCP/IP and provides a printer session
emulating an IBM 3812 printer.  The printer session works with
.B scs2ascii
and host print transform on the AS/400 to print AS/400 spooled files
on a local printer.
.PP
To end a printer session, send it a
.I SIGTERM
signal with
.BR kill (1).
.PP
Printing support requires named session support, which is available
for most recent versions of OS/400, but is only supplied with V4R3 and
later versions.  For V3R2, V3R7, V4R1 and V4R2, please see informational APAR
II10918.
.SH OPTIONS
Additional
.B lp5250d
options are described in the
.BR tn5250rc (5)
man page.
.TP
\fB\-u\fR, \fB\-\-user=\fIUSER\fR
run lp5250d as \fIUSER
.TP
\fB\-H\fR, \fB\-\-help\fR
display this help and exit
.TP
\fB\-v\fR, \fB\-\-version\fR
output version information and exit
.SH EXAMPLES
.TP
.I "lp5250d env.DEVNAME=LINPRT1 env.IBMMFRTYPMDL='*HP4' as400sys"
Connect to
.I as400sys
as the printer
.IR LINPRT1 ,
using host print transform on the AS/400 to print to an HP Laserjet or
compatible using lpr.
.TP
.I "lp5250d env.DEVNAME=LINPRT1 env.IBMMFRTYPMDL='*OKI184IBM' outputcommand=''scs2ascii | lpr -Poki' as400sys"
Print to an Okidata Microline 184 Turbo on the local print queue named
.RB ` oki '.
.TP
.I "lp5250d env.DEVNAME=PSPRT outputcommand='scs2ps > scs$$.ps' as400sys"
Convert spooled files into PostScript\*R files.
.SH BUGS
Please report any bugs you find to https://github.com/tn5250/tn5250/issues
.SH "SEE ALSO"
.BR tn5250rc (5),
.BR tn5250 (1),
.BR scs2ascii (1),
.BR scs2ps (1),
.BR https://tn5250.github.io/ ,
.BR RFC1205 ,
.BR RFC2877 ,
.BR II10918 .
.SH COPYRIGHT
.B tn5250
is copyright
.if t \(co
1997 \- 2001 Michael Madore.  This manpage is copyright
.if t \(co
1999 \- 2001 Carey Evans.
.PP
This program is free software; you can redistribute it and/or modify
it under the terms of the GNU Lesser General Public License as published by
the Free Software Foundation; either version 2.1 of the License, or
(at your option) any later version.
.PP
This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU Lesser General Public License for more details.
.PP
You should have received a copy of the GNU Lesser General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
.SH TRADEMARKS
PostScript is a trademark of Adobe Systems Incorporated.
.SH AUTHORS
.B tn5250
was written by Michael Madore, Jay Felice, Scott Klement
and others; see the AUTHORS file for details.
This manual page was written by Carey Evans.


NEWS
tn5250 NEWS
------------------------------------------------------------------------------

Wed May 04, 2005:
	* The tn5250 project is hosted on sourceforge at:
		http://tn5250.sourceforge.net
	* There is a mailing list and archives at David Gibbs midrange site:
		http://www.midrange.com
	  The mailing list is at:
		http://lists.midrange.com/mailman/listinfo/linux5250
	* There is an X11 front end written by James Rich.  The site is:
		http://www.chowhouse.com/~james/x5250
	* For all other news please visit the project page on sourceforge.

Mon Dec 06, 1999:
	* A FAQ is now in existence, thanks to Philippe Petrinko.  You can
	  browse the FAQ at http://perso.libertysurf.fr/plinux/tn5250-faq.html
	* The CVS server is now serving up the mailing list archives, thanks
	  to David Gibbs.   You can browse the archives at
	  http://cvs.cleveland.lug.net/mail-archives/tn5250



lp5250d/lp5250d.c
/* TN5250 - An implementation of the 5250 telnet protocol.
 * Copyright (C) 1997-2008 Michael Madore
 * 
 * This file is part of TN5250.
 *
 * TN5250 is free software; you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation; either version 2.1, or (at your option)
 * any later version.
 * 
 * TN5250 is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 * 
 * You should have received a copy of the GNU Lesser General Public License
 * along with this software; see the file COPYING.  If not, write to
 * the Free Software Foundation, Inc., 59 Temple Place, Suite 330,
 * Boston, MA 02111-1307 USA
 * 
 */

#include "tn5250-private.h"
#include <syslog.h>

/* If getopt.h exists then getopt_long() probably does as well.  If
 * getopt.h doesn't exist (like on Solaris) then we probably need to use
 * plain old getopt().
 */
#ifdef HAVE_GETOPT_H
#define _GNU_SOURCE
#include <getopt.h>
#endif

static void syntax (void);

/* This just checks what arguments (options) were passed to lp5250d on the
 * command line.
 */
int check_options (int argc, char **argv, char **user, int *nodaemon);

/* This sets the current UID to the uid of the user passed.
 */
int change_user (char *user);

extern char *version_string;

Tn5250PrintSession *printsess = NULL;
Tn5250Stream *stream = NULL;
Tn5250Config *config = NULL;

int
main (int argc, char *argv[])
{
  char *user = NULL;
  int nodaemon = 0;

  if (check_options (argc, argv, &user, &nodaemon) != 0)
    {
      exit (1);
    }

  /* Change the UID if required before parsing the options */
  if (user != NULL)
    {
      if (change_user (user) != 0)
	{
	  exit (1);
	}
    }

  config = tn5250_config_new ();
  if (tn5250_config_load_default (config) == -1)
    {
      tn5250_config_unref (config);
      exit (1);
    }

  if (tn5250_config_parse_argv (config, argc, argv) == -1)
    {
      tn5250_config_unref (config);
      syntax ();
    }

  if (tn5250_config_get (config, "help"))
    {
      syntax ();
    }
  else if (tn5250_config_get (config, "version"))
    {
      printf ("tn5250 version %s\n", version_string);
      exit (0);
    }
  else if (!tn5250_config_get (config, "host"))
    {
      syntax ();
    }

  if (!nodaemon)
      nodaemon = tn5250_config_get_bool (config, "nodaemon");

  if (!nodaemon) 
      if (tn5250_daemon (0, 0, 0) < 0)
        {
          perror ("tn5250_daemon");
          exit (2);
        }

#ifndef NDEBUG
  if (tn5250_config_get (config, "trace"))
    {
      tn5250_log_open (tn5250_config_get (config, "trace"));
      TN5250_LOG (("lp5250d version %s, built on %s\n", version_string,
		   __DATE__));
      TN5250_LOG (("host = %s\n", tn5250_config_get (config, "host")));
    }
#endif

  openlog ("lp5250d", LOG_PID, LOG_DAEMON);

  stream = tn5250_stream_open (tn5250_config_get (config, "host"), config);
  if (stream == NULL)
    {
      syslog (LOG_INFO, "Couldn't connect to %s",
	      tn5250_config_get (config, "host"));
      exit (1);
    }

  printsess = tn5250_print_session_new ();
  tn5250_stream_setenv (stream, "TERM", "IBM-3812-1");
/*    tn5250_stream_setenv(stream, "DEVNAME", sessionname); */
  tn5250_stream_setenv (stream, "IBMFONT", "12");
  syslog (LOG_INFO, "DEVNAME = %s", tn5250_stream_getenv (stream, "DEVNAME"));
  if (tn5250_stream_getenv (stream, "IBMMFRTYPMDL"))
    {
      syslog (LOG_INFO, "TRANSFORM = %s",
	      tn5250_stream_getenv (stream, "IBMMFRTYPMDL"));
      tn5250_stream_setenv (stream, "IBMTRANSFORM", "1");
    }
  else
    {
      tn5250_stream_setenv (stream, "IBMTRANSFORM", "0");
    }
  tn5250_print_session_set_fd (printsess,
			       tn5250_stream_socket_handle (stream));
  tn5250_print_session_set_stream (printsess, stream);
  if (tn5250_config_get (config, "map"))
    {
      tn5250_print_session_set_char_map (printsess,
					 tn5250_config_get (config, "map"));
      tn5250_setenv ("TN5250_CCSIDMAP", tn5250_config_get (config, "map"), 0);
    }
  else
    {
      tn5250_print_session_set_char_map (printsess, "37");
      tn5250_setenv ("TN5250_CCSIDMAP", "37", 0);
    }
  if (tn5250_config_get (config, "outputcommand"))
    {
      tn5250_print_session_set_output_command (printsess,
					       tn5250_config_get (config,
								  "outputcommand"));
    }
  else
    {
      tn5250_print_session_set_output_command (printsess, "scs2ascii|lpr");
    }

  tn5250_print_session_main_loop (printsess);

  tn5250_print_session_destroy (printsess);
  tn5250_stream_destroy (stream);
  if (config != NULL)
    {
      tn5250_config_unref (config);
    }
#ifndef NDEBUG
  tn5250_log_close ();
#endif
  return 0;
}

static void
syntax ()
{
  printf ("Usage:  lp5250d [options] host[:port]\n"
	  "Options:\n"
	  "\ttrace=FILE                 specify FULL path to log file\n"
	  "\tmap=NAME                   specify translation map\n"
	  "\tenv.DEVNAME=NAME           specify session name\n"
	  "\tenv.IBMMFRTYPMDL=NAME      specify host print transform name\n"
	  "\toutputcommand=CMD          specify the print output command\n"
	  "\t-u,--user=NAME             display user to run as\n"
	  "\t-v,--version               display version\n"
	  "\t-N,--nodaemon              do not run as daemon (run in foreground)\n"
	  "\t-H,--help                  display this help\n");

  exit (255);
}


/* This checks the options lp5250d was passed.
 */
int
check_options (int argc, char **argv, char **user, int *nodaemon)
{
#ifdef HAVE_GETOPT_H
  struct option options[4];
#endif
  int i;
  extern char *optarg;

#ifdef HAVE_GETOPT_H
  options[0].name = "user";
  options[0].has_arg = required_argument;
  options[0].flag = NULL;
  options[0].val = 'u';
  options[1].name = "version";
  options[1].has_arg = no_argument;
  options[1].flag = NULL;
  options[1].val = 'v';
  options[2].name = "help";
  options[2].has_arg = no_argument;
  options[2].flag = NULL;
  options[2].val = 'H';
  options[3].name = "nodaemon";
  options[3].has_arg = no_argument;
  options[3].flag = NULL;
  options[3].val = 'N';
  options[4].name = 0;
  options[4].has_arg = 0;
  options[4].flag = 0;
  options[4].val = 0;
#endif

#ifdef HAVE_GETOPT_H
  while ((i = getopt_long (argc, argv, "u:vHN", options, NULL)) != -1)
#else
  while ((i = getopt (argc, argv, "u:vHN")) != -1)
#endif
    {
      switch (i)
	{
	case '?':
	  syntax ();
	  return -1;
	case 'u':
	  *user = optarg;
	  break;
	case 'H':
	  syntax ();
	  return -1;
	case 'v':
	  printf ("lp5250d version:  %s\n", VERSION);
	  return -1;
	case 'N':
	  *nodaemon=1;
          break;
	default:
	  syntax ();
	  return -1;
	}
    }
  return 0;
}


/* Set the UID to run lp5250d as.
 */
int
change_user (char *user)
{
  struct passwd *pwent;

  errno = 0;
  pwent = getpwnam (user);

  if (pwent == NULL)
    {
      printf ("Unable to set UID to user %s\n", user);
      perror ("getpwnam()");
      return (-1);
    }

  if (setuid (pwent->pw_uid) != 0)
    {
      printf ("Unable to set UID to user %s\n", user);
      perror ("setuid()");
      return (-1);
    }

  return (0);
}
